File: parcial/src/main/java/com/examen/nodes/nodo_2/nodo.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
	"strconv"
)

type Nodo struct {
	nodeId        string
	portNode      int
	serverHost    string
	serverPort    int
	serverConn    net.Conn
	listener      net.Listener
	rootDirectory string
	tableCounts   []string
	fileLock      sync.Mutex
	transactionLock sync.Mutex
	transactionId int
}

func NewNodo(nodeId string, portNode int, rootDirectory string) *Nodo {
	n := &Nodo{
		nodeId:        nodeId,
		portNode:      portNode,
		serverHost:    "127.0.0.1",
		serverPort:    5000,
		rootDirectory: rootDirectory,
		tableCounts:   []string{"cu_1.txt", "cu_2.txt", "cu_3.txt"},
	}
	if !strings.HasSuffix(rootDirectory, string(filepath.Separator)) {
		n.rootDirectory += string(filepath.Separator)
	}
	n.initializeTransactionFile()
	return n
}

func (n *Nodo) initializeTransactionFile() {
	transactionFile := filepath.Join(n.rootDirectory, "transacciones.txt")
	if _, err := os.Stat(transactionFile); os.IsNotExist(err) {
		file, err := os.Create(transactionFile)
		if err != nil {
			fmt.Printf("Error al crear transacciones.txt: %v\n", err)
			return
		}
		defer file.Close()
		writer := bufio.NewWriter(file)
		writer.WriteString("ID_TRANSACC | ID_ORIG | ID_DEST | MONTO  | FECHA_HORA         | ESTADO\n")
		writer.WriteString("----------------------------------------------------------------------------\n")
		writer.Flush()
		now := time.Now().Format("15:04:05")
		fmt.Printf("Archivo transacciones.txt creado con encabezado a las %s\n", now)
	} else {
		file, err := os.Open(transactionFile)
		if err != nil {
			fmt.Printf("Error al leer transacciones.txt: %v\n", err)
			return
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		scanner.Scan() // Saltar encabezado
		scanner.Scan() // Saltar línea divisoria
		for scanner.Scan() {
			line := scanner.Text()
			tokens := strings.Split(line, "|")
			if len(tokens) >= 1 {
				if id, err := strconv.Atoi(strings.TrimSpace(tokens[0])); err == nil {
					n.transactionId = id
				} else {
					fmt.Printf("Error al parsear ID_TRANSACC en transacciones.txt: %s\n", tokens[0])
				}
			}
		}
		now := time.Now().Format("15:04:05")
		fmt.Printf("Archivo transacciones.txt existente. Último ID_TRANSACC leído: %d a las %s\n", n.transactionId, now)
	}
}

func (n *Nodo) Run() {
	err := n.connectionServer()
	if err != nil {
		fmt.Printf("Error al crear el nodo %s: %v\n", n.nodeId, err)
		return
	}
	n.registerNode()
	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", n.portNode))
	if err != nil {
		fmt.Printf("Error al crear el nodo %s: %v\n", n.nodeId, err)
		return
	}
	n.listener = listener
	now := time.Now().Format("15:04:05")
	fmt.Printf("Nodo %s iniciado en el puerto %d a las %s\n", n.nodeId, n.portNode, now)

	go n.handleTasks()
	go n.handleUpdates()

	select {} // Mantener el programa corriendo
}

func (n *Nodo) connectionServer() error {
	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", n.serverHost, n.serverPort))
	if err != nil {
		return err
	}
	n.serverConn = conn
	return nil
}

func (n *Nodo) registerNode() error {
	ip := getLocalIP() // Obtener la IP del nodo
	tables := strings.Join(n.tableCounts, ",")
	msg := fmt.Sprintf("REGISTRO_NODO:%s:%s:%d:%s", n.nodeId, ip, n.portNode, tables)
	fmt.Fprintf(n.serverConn, "%s\n", msg)
	now := time.Now().Format("15:04:05")
	fmt.Printf("Nodo %s registrado en el servidor con IP %s y tablas: %s a las %s\n", n.nodeId, ip, tables, now)
	return nil
}

func getLocalIP() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		return "127.0.0.1"
	}
	for _, addr := range addrs {
		if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
			if ipnet.IP.To4() != nil {
				return ipnet.IP.String()
			}
		}
	}
	return "127.0.0.1"
}

func (n *Nodo) handleTasks() {
	for {
		conn, err := n.listener.Accept()
		if err != nil {
			fmt.Printf("Error al aceptar tareas en %s: %v\n", n.nodeId, err)
			continue
		}
		now := time.Now().Format("15:04:05")
		fmt.Printf("Nodo %s recibió una tarea de %s a las %s\n", n.nodeId, conn.RemoteAddr(), now)
		go n.handleTask(conn)
	}
}

func (n *Nodo) handleTask(conn net.Conn) {
	defer conn.Close()
	reader := bufio.NewReader(conn)
	writer := bufio.NewWriter(conn)

	option, err := reader.ReadString('\n')
	if err != nil {
		fmt.Printf("Error al leer tarea en %s: %v\n", n.nodeId, err)
		return
	}
	option = strings.TrimSpace(option)

	if strings.HasPrefix(option, "1") {
		idCount := strings.Split(option, "-")[1]
		result := n.checkBalance(idCount)
		writer.WriteString(fmt.Sprintf("RESULTADO %s -> %s\n", n.nodeId, result))
	} else if strings.HasPrefix(option, "2") {
		tokens := strings.Split(option, "-")
		details := strings.Split(tokens[1], ":")
		idOrigen := details[0]
		idDestino := details[1]
		montoStr := details[2]
		result := n.transferBalance(idOrigen, idDestino, montoStr)
		writer.WriteString(fmt.Sprintf("RESULTADO %s -> TRANSFERENCIA DE %s A %s POR %s - %s\n", n.nodeId, idOrigen, idDestino, montoStr, result))
	}
	writer.Flush()
}

func (n *Nodo) checkBalance(idClient string) string {
	for _, fileName := range n.tableCounts {
		filePath := filepath.Join(n.rootDirectory, fileName)
		if _, err := os.Stat(filePath); os.IsNotExist(err) {
			return fmt.Sprintf("Archivo no encontrado: %s", fileName)
		}
		now := time.Now().Format("15:04:05")
		fmt.Printf("Leyendo archivo para checkBalance: %s a las %s\n", filePath, now)
		file, err := os.Open(filePath)
		if err != nil {
			return fmt.Sprintf("Error al leer el nodo %s: %v", n.nodeId, err)
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		scanner.Scan() // Saltar encabezado
		scanner.Scan() // Saltar línea divisoria
		for scanner.Scan() {
			line := scanner.Text()
			tokens := strings.Split(line, "|")
			if len(tokens) >= 3 {
				idCount := strings.TrimSpace(tokens[0])
				if idCount == idClient {
					saldoStr := strings.TrimSpace(strings.ReplaceAll(tokens[2], ",", ""))
					return fmt.Sprintf("SALDO: %s", saldoStr)
				}
			}
		}
	}
	return "CUENTA NO ENCONTRADA"
}

func (n *Nodo) logTransaction(idOrigen, idDestino string, monto float64, estado string) {
	n.transactionLock.Lock()
	defer n.transactionLock.Unlock()
	n.transactionId++
	now := time.Now().Format("2006-01-02 15:04:05")
	line := fmt.Sprintf("%-10d | %-7s | %-7s | %-7.2f | %-19s | %s\n",
		n.transactionId, idOrigen, idDestino, monto, now, estado)
	transactionFile := filepath.Join(n.rootDirectory, "transacciones.txt")
	file, err := os.OpenFile(transactionFile, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Printf("Error al registrar la transacción: %v a las %s\n", err, time.Now().Format("15:04:05"))
		return
	}
	defer file.Close()
	writer := bufio.NewWriter(file)
	writer.WriteString(line)
	writer.Flush()
	nowStr := time.Now().Format("15:04:05")
	fmt.Printf("Transacción registrada y escrita: ID_TRANSACC=%d, ID_ORIG=%s, ID_DEST=%s, MONTO=%.2f, FECHA_HORA=%s, ESTADO=%s a las %s\n",
		n.transactionId, idOrigen, idDestino, monto, now, estado, nowStr)
}

func (n *Nodo) transferBalance(idOrigen, idDestino, amount string) string {
	now := time.Now().Format("15:04:05")
	fmt.Printf("TRANSFERENCIA: %s %s %s a las %s\n", idOrigen, idDestino, amount, now)
	monto, err := strconv.ParseFloat(amount, 64)
	if err != nil {
		n.logTransaction(idOrigen, idDestino, 0.0, "Pendiente")
		return fmt.Sprintf("Error al parsear el monto: %v", err)
	}

	var origenFile, destinoFile string
	var saldoOrigen, saldoDestino float64

	for _, fileName := range n.tableCounts {
		filePath := filepath.Join(n.rootDirectory, fileName)
		if _, err := os.Stat(filePath); os.IsNotExist(err) {
			n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
			return fmt.Sprintf("Archivo no encontrado: %s", fileName)
		}
		now := time.Now().Format("15:04:05")
		fmt.Printf("Buscando cuenta de origen en: %s a las %s\n", filePath, now)
		file, err := os.Open(filePath)
		if err != nil {
			n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
			return fmt.Sprintf("Error al leer el nodo %s: %v", n.nodeId, err)
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		scanner.Scan() // Saltar encabezado
		scanner.Scan() // Saltar línea divisoria
		for scanner.Scan() {
			line := scanner.Text()
			tokens := strings.Split(line, "|")
			if len(tokens) >= 3 {
				idCount := strings.TrimSpace(tokens[0])
				if idCount == idOrigen {
					saldoStr := strings.TrimSpace(strings.ReplaceAll(tokens[2], ",", ""))
					saldoOrigen, err = strconv.ParseFloat(saldoStr, 64)
					if err != nil {
						n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
						return fmt.Sprintf("Error al parsear el saldo de la cuenta de origen: %v", err)
					}
					fmt.Printf("Saldo de origen (%s): %f a las %s\n", idCount, saldoOrigen, now)
					origenFile = fileName
					break
				}
			}
		}
		if origenFile != "" {
			break
		}
	}
	if origenFile == "" {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return "Cuenta de origen no encontrada"
	}

	for _, fileName := range n.tableCounts {
		filePath := filepath.Join(n.rootDirectory, fileName)
		if _, err := os.Stat(filePath); os.IsNotExist(err) {
			n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
			return fmt.Sprintf("Archivo no encontrado: %s", fileName)
		}
		now := time.Now().Format("15:04:05")
		fmt.Printf("Buscando cuenta de destino en: %s a las %s\n", filePath, now)
		file, err := os.Open(filePath)
		if err != nil {
			n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
			return fmt.Sprintf("Error al leer el nodo %s: %v", n.nodeId, err)
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		scanner.Scan() // Saltar encabezado
		scanner.Scan() // Saltar línea divisoria
		for scanner.Scan() {
			line := scanner.Text()
			tokens := strings.Split(line, "|")
			if len(tokens) >= 3 {
				idCount := strings.TrimSpace(tokens[0])
				if idCount == idDestino {
					saldoStr := strings.TrimSpace(strings.ReplaceAll(tokens[2], ",", ""))
					saldoDestino, err = strconv.ParseFloat(saldoStr, 64)
					if err != nil {
						n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
						return fmt.Sprintf("Error al parsear el saldo de la cuenta de destino: %v", err)
					}
					fmt.Printf("Saldo de destino (%s): %f a las %s\n", idCount, saldoDestino, now)
					destinoFile = fileName
					break
				}
			}
		}
		if destinoFile != "" {
			break
		}
	}
	if destinoFile == "" {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return "Cuenta de destino no encontrada"
	}

	if saldoOrigen < monto {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return "Saldo insuficiente en la cuenta de origen"
	}

	saldoOrigen -= monto
	saldoDestino += monto

	n.fileLock.Lock()
	defer n.fileLock.Unlock()

	// Actualizar archivo de origen
	lines, err := readFileLines(filepath.Join(n.rootDirectory, origenFile))
	if err != nil {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return fmt.Sprintf("Error al leer el archivo de origen: %v", err)
	}
	err = writeFileLines(filepath.Join(n.rootDirectory, origenFile), updateLines(lines, idOrigen, saldoOrigen))
	if err != nil {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return fmt.Sprintf("Error al actualizar la cuenta de origen: %v", err)
	}
	now = time.Now().Format("15:04:05")
	fmt.Printf("Archivo de origen %s actualizado con saldo: %.2f a las %s\n", origenFile, saldoOrigen, now)

	// Enviar actualización del archivo origen
	content, err := readFileContent(filepath.Join(n.rootDirectory, origenFile))
	if err == nil {
		fmt.Fprintf(n.serverConn, "UPDATE:%s:%s:%s\n", n.nodeId, origenFile, strings.ReplaceAll(content, "\n", "\\n"))
	}

	// Actualizar archivo de destino
	lines, err = readFileLines(filepath.Join(n.rootDirectory, destinoFile))
	if err != nil {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return fmt.Sprintf("Error al leer el archivo de destino: %v", err)
	}
	err = writeFileLines(filepath.Join(n.rootDirectory, destinoFile), updateLines(lines, idDestino, saldoDestino))
	if err != nil {
		n.logTransaction(idOrigen, idDestino, monto, "Pendiente")
		return fmt.Sprintf("Error al actualizar la cuenta de destino: %v", err)
	}
	now = time.Now().Format("15:04:05")
	fmt.Printf("Archivo de destino %s actualizado con saldo: %.2f a las %s\n", destinoFile, saldoDestino, now)

	// Enviar actualización del archivo destino
	content, err = readFileContent(filepath.Join(n.rootDirectory, destinoFile))
	if err == nil {
		fmt.Fprintf(n.serverConn, "UPDATE:%s:%s:%s\n", n.nodeId, destinoFile, strings.ReplaceAll(content, "\n", "\\n"))
	}

	n.logTransaction(idOrigen, idDestino, monto, "Confirmada")
	return fmt.Sprintf("Transferencia exitosa en %s y %s. Saldo origen: %.2f, Saldo destino: %.2f", origenFile, destinoFile, saldoOrigen, saldoDestino)
}

func readFileLines(filePath string) ([]string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func writeFileLines(filePath string, lines []string) error {
	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer file.Close()
	writer := bufio.NewWriter(file)
	for _, line := range lines {
		writer.WriteString(line + "\n")
	}
	return writer.Flush()
}

func updateLines(lines []string, id string, saldo float64) []string {
	for i := 2; i < len(lines); i++ {
		tokens := strings.Split(lines[i], "|")
		if len(tokens) >= 3 {
			idCount := strings.TrimSpace(tokens[0])
			if idCount == id {
				lines[i] = fmt.Sprintf("%s | %s | %.2f | %s", idCount, strings.TrimSpace(tokens[1]), saldo, strings.TrimSpace(tokens[3]))
			}
		}
	}
	return lines
}

func readFileContent(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()
	var content strings.Builder
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		content.WriteString(scanner.Text() + "\n")
	}
	return content.String(), scanner.Err()
}

func (n *Nodo) handleUpdates() {
	reader := bufio.NewReader(n.serverConn)
	for {
		update, err := reader.ReadString('\n')
		if err != nil {
			fmt.Printf("Error al manejar actualizaciones en %s: %v\n", n.nodeId, err)
			return
		}
		update = strings.TrimSpace(update)
		if strings.HasPrefix(update, "UPDATE:") {
			now := time.Now().Format("15:04:05")
			fmt.Printf("%s recibió actualización: %s a las %s\n", n.nodeId, update, now)
			parts := strings.SplitN(update, ":", 4)
			if len(parts) >= 4 {
				tabla := parts[2]
				contenido := strings.ReplaceAll(parts[3], "\\n", "\n")
				if contains(n.tableCounts, tabla) {
					n.aplicarActualizacion(tabla, contenido)
				}
			}
		}
	}
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func (n *Nodo) aplicarActualizacion(tabla, contenido string) {
	n.fileLock.Lock()
	defer n.fileLock.Unlock()
	filePath := filepath.Join(n.rootDirectory, tabla)
	file, err := os.Create(filePath)
	if err != nil {
		fmt.Printf("Error al reemplazar el contenido de %s en %s: %v\n", tabla, n.nodeId, err)
		return
	}
	defer file.Close()
	writer := bufio.NewWriter(file)
	writer.WriteString(contenido)
	writer.Flush()
	now := time.Now().Format("15:04:05")
	fmt.Printf("%s reemplazó el contenido de %s a las %s\n", n.nodeId, tabla, now)
}

func main() {
	rootDirectory := "C:\\Users\\sergi\\transacciones_bancarias\\parcial\\src\\main\\java\\com\\examen\\nodes\\nodo_1\\"
	n := NewNodo("nodo_1", 6001, rootDirectory)
	go n.Run()
	time.Sleep(time.Hour * 24) // Mantener el programa vivo (ajustar según necesidad)
}

File: parcial/src/main/java/com/examen/nodes/nodo_3/nodo.rb
#!/usr/bin/env ruby
require 'socket'
require 'thread'
require 'fileutils'
require 'time'

class Nodo
  attr_reader :node_id, :port_node, :root_directory, :table_counts

  def initialize(node_id, port_node, root_directory)
    @node_id = node_id
    @port_node = port_node
    @server_host = ' ' # Colocar el IP del servidor
    @server_port = 5000
    @root_directory = root_directory.end_with?(File::SEPARATOR) ? root_directory : "#{root_directory}#{File::SEPARATOR}"
    @table_counts = ['cu_1.txt', 'cu_2.txt', 'cu_3.txt']
    @file_lock = Mutex.new
    @transaction_lock = Mutex.new
    @transaction_id = 0
    initialize_transaction_file
  end

  def initialize_transaction_file
    transaction_file = File.join(@root_directory, 'transacciones.txt')
    
    if !File.exist?(transaction_file)
      begin
        File.open(transaction_file, 'w') do |file|
          file.puts "ID_TRANSACC | ID_ORIG | ID_DEST | MONTO  | FECHA_HORA         | ESTADO"
          file.puts "----------------------------------------------------------------------------"
        end
        now = Time.now.strftime("%H:%M:%S")
        puts "Archivo transacciones.txt creado con encabezado a las #{now}"
      rescue => e
        puts "Error al crear transacciones.txt: #{e.message}"
      end
    else
      begin
        File.open(transaction_file, 'r') do |file|
          file.readline  # Saltar encabezado
          file.readline  # Saltar línea divisoria
          
          file.each_line do |line|
            tokens = line.split('|')
            if tokens.length >= 1
              begin
                @transaction_id = tokens[0].strip.to_i
              rescue => e
                puts "Error al parsear ID_TRANSACC en transacciones.txt: #{tokens[0].strip}"
              end
            end
          end
        end
        now = Time.now.strftime("%H:%M:%S")
        puts "Archivo transacciones.txt existente. Último ID_TRANSACC leído: #{@transaction_id} a las #{now}"
      rescue => e
        puts "Error al leer transacciones.txt: #{e.message}"
      end
    end
  end

  def run
    begin
      connection_server
      register_node
      @listener = TCPServer.new(@port_node)
      now = Time.now.strftime("%H:%M:%S")
      puts "Nodo #{@node_id} iniciado en el puerto #{@port_node} a las #{now}"

      Thread.new { handle_tasks }
      Thread.new { handle_updates }

      # Mantener el programa en ejecución
      loop do
        sleep 60
      end
    rescue => e
      puts "Error al crear el nodo #{@node_id}: #{e.message}"
    end
  end

  def connection_server
    @server_conn = TCPSocket.new(@server_host, @server_port)
  end

  def register_node
    ip = get_local_ip
    tables = @table_counts.join(',')
    msg = "REGISTRO_NODO:#{@node_id}:#{ip}:#{@port_node}:#{tables}"
    @server_conn.puts(msg)
    now = Time.now.strftime("%H:%M:%S")
    puts "Nodo #{@node_id} registrado en el servidor con IP #{ip} y tablas: #{tables} a las #{now}"
  end

  def get_local_ip
    # Intenta obtener la IP local
    begin
      # Crea un socket UDPSocket temporal
      socket = UDPSocket.new
      socket.connect("8.8.8.8", 1)  # No se envía ningún dato
      socket.addr.last
    rescue
      "127.0.0.1"
    ensure
      socket.close if socket
    end
  end

  def handle_tasks
    loop do
      begin
        client = @listener.accept
        client_addr = client.peeraddr[3]
        now = Time.now.strftime("%H:%M:%S")
        puts "Nodo #{@node_id} recibió una tarea de #{client_addr} a las #{now}"
        
        Thread.new { handle_task(client) }
      rescue => e
        puts "Error al aceptar tareas en #{@node_id}: #{e.message}"
      end
    end
  end

  def handle_task(client)
    begin
      option = client.gets.strip
      
      if option.start_with?('1')
        id_count = option.split('-')[1]
        result = check_balance(id_count)
        client.puts("RESULTADO #{@node_id} -> #{result}")
      elsif option.start_with?('2')
        tokens = option.split('-')
        details = tokens[1].split(':')
        id_origen = details[0]
        id_destino = details[1]
        monto_str = details[2]
        result = transfer_balance(id_origen, id_destino, monto_str)
        client.puts("RESULTADO #{@node_id} -> TRANSFERENCIA DE #{id_origen} A #{id_destino} POR #{monto_str} - #{result}")
      end
    rescue => e
      puts "Error al manejar la tarea en #{@node_id}: #{e.message}"
    ensure
      client.close
    end
  end

  def check_balance(id_client)
    @table_counts.each do |file_name|
      begin
        file_path = File.join(@root_directory, file_name)
        unless File.exist?(file_path)
          return "Archivo no encontrado: #{file_name}"
        end
        
        now = Time.now.strftime("%H:%M:%S")
        puts "Leyendo archivo para checkBalance: #{file_path} a las #{now}"
        
        File.open(file_path, 'r') do |file|
          file.readline  # Saltar encabezado
          file.readline  # Saltar línea divisoria
          
          file.each_line do |line|
            tokens = line.split('|')
            if tokens.length >= 3
              id_count = tokens[0].strip
              if id_count == id_client
                saldo_str = tokens[2].strip.gsub(',', '')
                return "SALDO: #{saldo_str}"
              end
            end
          end
        end
      rescue => e
        return "Error al leer el nodo #{@node_id}: #{e.message}"
      end
    end
    return "CUENTA NO ENCONTRADA"
  end

  def log_transaction(id_origen, id_destino, monto, estado)
    @transaction_lock.synchronize do
      @transaction_id += 1
      now = Time.now
      fecha_hora = now.strftime("%Y-%m-%d %H:%M:%S")
      
      begin
        transaction_file = File.join(@root_directory, 'transacciones.txt')
        File.open(transaction_file, 'a') do |file|
          line = sprintf("%-10d | %-7s | %-7s | %-7.2f | %-19s | %s\n", 
                         @transaction_id, id_origen, id_destino, monto, fecha_hora, estado)
          file.write(line)
        end
        now_str = Time.now.strftime("%H:%M:%S")
        puts "Transacción registrada y escrita: ID_TRANSACC=#{@transaction_id}, ID_ORIG=#{id_origen}, " +
             "ID_DEST=#{id_destino}, MONTO=#{monto}, FECHA_HORA=#{fecha_hora}, ESTADO=#{estado} a las #{now_str}"
      rescue => e
        now_str = Time.now.strftime("%H:%M:%S")
        puts "Error al registrar la transacción: #{e.message} a las #{now_str}"
      end
    end
  end

  def transfer_balance(id_origen, id_destino, amount_str)
    now = Time.now.strftime("%H:%M:%S")
    puts "TRANSFERENCIA: #{id_origen} #{id_destino} #{amount_str} a las #{now}"
    
    begin
      monto = Float(amount_str)
    rescue => e
      log_transaction(id_origen, id_destino, 0.0, "Pendiente")
      return "Error al parsear el monto: #{e.message}"
    end

    origen_file = nil
    destino_file = nil
    saldo_origen = 0.0
    saldo_destino = 0.0

    # Buscar cuenta de origen
    @table_counts.each do |file_name|
      begin
        file_path = File.join(@root_directory, file_name)
        unless File.exist?(file_path)
          log_transaction(id_origen, id_destino, monto, "Pendiente")
          return "Archivo no encontrado: #{file_name}"
        end
        
        now = Time.now.strftime("%H:%M:%S")
        puts "Buscando cuenta de origen en: #{file_path} a las #{now}"
        
        File.open(file_path, 'r') do |file|
          file.readline  # Saltar encabezado
          file.readline  # Saltar línea divisoria
          
          file.each_line do |line|
            tokens = line.split('|')
            if tokens.length >= 3
              id_count = tokens[0].strip
              if id_count == id_origen
                saldo_str = tokens[2].strip.gsub(',', '')
                saldo_origen = Float(saldo_str)
                origen_file = file_name
                puts "Saldo de origen (#{id_count}): #{saldo_origen} a las #{now}"
                break
              end
            end
          end
        end
        
        break if origen_file
      rescue => e
        log_transaction(id_origen, id_destino, monto, "Pendiente")
        return "Error al leer el nodo #{@node_id}: #{e.message}"
      end
    end

    if origen_file.nil?
      log_transaction(id_origen, id_destino, monto, "Pendiente")
      return "Cuenta de origen no encontrada"
    end

    # Buscar cuenta de destino
    @table_counts.each do |file_name|
      begin
        file_path = File.join(@root_directory, file_name)
        unless File.exist?(file_path)
          log_transaction(id_origen, id_destino, monto, "Pendiente")
          return "Archivo no encontrado: #{file_name}"
        end
        
        now = Time.now.strftime("%H:%M:%S")
        puts "Buscando cuenta de destino en: #{file_path} a las #{now}"
        
        File.open(file_path, 'r') do |file|
          file.readline  # Saltar encabezado
          file.readline  # Saltar línea divisoria
          
          file.each_line do |line|
            tokens = line.split('|')
            if tokens.length >= 3
              id_count = tokens[0].strip
              if id_count == id_destino
                saldo_str = tokens[2].strip.gsub(',', '')
                saldo_destino = Float(saldo_str)
                destino_file = file_name
                puts "Saldo de destino (#{id_count}): #{saldo_destino} a las #{now}"
                break
              end
            end
          end
        end
        
        break if destino_file
      rescue => e
        log_transaction(id_origen, id_destino, monto, "Pendiente")
        return "Error al leer el nodo #{@node_id}: #{e.message}"
      end
    end

    if destino_file.nil?
      log_transaction(id_origen, id_destino, monto, "Pendiente")
      return "Cuenta de destino no encontrada"
    end

    if saldo_origen < monto
      log_transaction(id_origen, id_destino, monto, "Pendiente")
      return "Saldo insuficiente en la cuenta de origen"
    end

    saldo_origen -= monto
    saldo_destino += monto

    @file_lock.synchronize do
      # Actualizar archivo de origen
      begin
        origen_path = File.join(@root_directory, origen_file)
        lines = File.readlines(origen_path)
        
        File.open(origen_path, 'w') do |file|
          file.write(lines[0])
          file.write(lines[1])
          
          lines[2..-1].each do |line|
            tokens = line.split('|')
            if tokens.length >= 3
              id_count = tokens[0].strip
              if id_count == id_origen
                file.puts("#{id_count} | #{tokens[1].strip} | #{sprintf('%.2f', saldo_origen)} | #{tokens[3].strip}")
              else
                file.write(line)
              end
            end
          end
        end
        
        now = Time.now.strftime("%H:%M:%S")
        puts "Archivo de origen #{origen_file} actualizado con saldo: #{sprintf('%.2f', saldo_origen)} a las #{now}"
        
        # Enviar actualización del archivo origen
        content = File.read(origen_path)
        @server_conn.puts("UPDATE:#{@node_id}:#{origen_file}:#{content.gsub("\n", "\\n")}")
      rescue => e
        log_transaction(id_origen, id_destino, monto, "Pendiente")
        return "Error al actualizar la cuenta de origen: #{e.message}"
      end

      # Actualizar archivo de destino
      begin
        destino_path = File.join(@root_directory, destino_file)
        lines = File.readlines(destino_path)
        
        File.open(destino_path, 'w') do |file|
          file.write(lines[0])
          file.write(lines[1])
          
          lines[2..-1].each do |line|
            tokens = line.split('|')
            if tokens.length >= 3
              id_count = tokens[0].strip
              if id_count == id_destino
                file.puts("#{id_count} | #{tokens[1].strip} | #{sprintf('%.2f', saldo_destino)} | #{tokens[3].strip}")
              else
                file.write(line)
              end
            end
          end
        end
        
        now = Time.now.strftime("%H:%M:%S")
        puts "Archivo de destino #{destino_file} actualizado con saldo: #{sprintf('%.2f', saldo_destino)} a las #{now}"
        
        # Enviar actualización del archivo destino
        content = File.read(destino_path)
        @server_conn.puts("UPDATE:#{@node_id}:#{destino_file}:#{content.gsub("\n", "\\n")}")
      rescue => e
        log_transaction(id_origen, id_destino, monto, "Pendiente")
        return "Error al actualizar la cuenta de destino: #{e.message}"
      end
    end

    log_transaction(id_origen, id_destino, monto, "Confirmada")
    return "Transferencia exitosa en #{origen_file} y #{destino_file}. Saldo origen: #{sprintf('%.2f', saldo_origen)}, Saldo destino: #{sprintf('%.2f', saldo_destino)}"
  end

  def handle_updates
    begin
      loop do
        update = @server_conn.gets.strip
        if update.start_with?('UPDATE:')
          now = Time.now.strftime("%H:%M:%S")
          puts "#{@node_id} recibió actualización: #{update} a las #{now}"
          
          parts = update.split(':', 4)
          if parts.length >= 4
            tabla = parts[2]
            contenido = parts[3].gsub('\\n', "\n")
            
            if @table_counts.include?(tabla)
              aplicar_actualizacion(tabla, contenido)
            end
          end
        end
      end
    rescue => e
      puts "Error al manejar actualizaciones en #{@node_id}: #{e.message}"
    end
  end

  def aplicar_actualizacion(tabla, contenido)
    @file_lock.synchronize do
      begin
        file_path = File.join(@root_directory, tabla)
        File.write(file_path, contenido)
        now = Time.now.strftime("%H:%M:%S")
        puts "#{@node_id} reemplazó el contenido de #{tabla} a las #{now}"
      rescue => e
        puts "Error al reemplazar el contenido de #{tabla} en #{@node_id}: #{e.message}"
      end
    end
  end
end

# Punto de entrada principal
if __FILE__ == $PROGRAM_NAME
  # Configuración del nodo - ajustar según necesidad
  root_directory = " " # Colocar la ruta de los directorios
  node = Nodo.new("nodo_ruby",  , root_directory) # Colocar el puerto del nodo
  node.run
end
